README
NAME
    Scalar::Does - like ref() but useful

SYNOPSIS
      use Scalar::Does qw( -constants );
  
      my $object = bless {}, 'Some::Class';
  
      does($object, 'Some::Class');   # true
      does($object, '%{}');           # true
      does($object, HASH);            # true
      does($object, ARRAY);           # false

DESCRIPTION
    It has long been noted that Perl would benefit from a `does()` built-in. A
    check that `ref($thing) eq 'ARRAY'` doesn't allow you to accept an object
    that uses overloading to provide an array-like interface.

  Functions
    `does($scalar, $role)`
        Checks if a scalar is capable of performing the given role. The
        following (case-sensitive) roles are predefined:

        *   SCALAR or ${}

            Checks if the scalar can be used as a scalar reference.

            Note: this role does not check whether a scalar is a scalar (which
            is obviously true) but whether it is a reference to another
            scalar.

        *   ARRAY or @{}

            Checks if the scalar can be used as an array reference.

        *   HASH or %{}

            Checks if the scalar can be used as a hash reference.

        *   CODE or &{}

            Checks if the scalar can be used as a code reference.

        *   GLOB or *{}

            Checks if the scalar can be used as a glob reference.

        *   REF

            Checks if the scalar can be used as a ref reference (i.e. a
            reference to another reference).

        *   LVALUE

            Checks if the scalar is a reference to a special lvalue (e.g. the
            result of `substr` or `splice`).

        *   IO or <>

            Uses IO::Detect to check if the scalar is a filehandle or
            file-handle-like object.

            (The `<>` check is slightly looser, allowing objects which
            overload `<>`, though overloading `<>` well can be a little
            tricky.)

        *   VSTRING

            Checks if the scalar is a vstring reference.

        *   FORMAT

            Checks if the scalar is a format reference.

        *   Regexp or qr

            Checks if the scalar can be used as a quoted regular expression.

        *   bool

            Checks if the scalar can be used as a boolean. (It's pretty rare
            for this to not be true.)

        *   ""

            Checks if the scalar can be used as a string. (It's pretty rare
            for this to not be true.)

        *   0+

            Checks if the scalar can be used as a number. (It's pretty rare
            for this to not be true.)

            Note that this is far looser than `looks_like_number` from
            Scalar::Util. For example, an unblessed arrayref can be used as a
            number (it numifies to its reference address); the string "Hello
            World" can be used as a number (it numifies to 0).

        *   ~~

            Checks if the scalar can be used on the right hand side of a smart
            match.

        If the given *role* is blessed, and provides a `check` method, then
        `does` delegates to that.

        Otherwise, if the scalar being tested is blessed, then
        `$scalar->DOES($role)` is called, and `does` returns true if the
        method call returned true.

        If the scalar being tested looks like a Perl class name, then
        `$scalar->DOES($role)` is also called, and the string "0E0" is
        returned for success, which evaluates to 0 in a numeric context but
        true in a boolean context.

    `does($role)`
        Called with a single argument, tests $_. Yes, this works with lexical
        $_.

          given ($object) {
             when(does ARRAY)  { ... }
             when(does HASH)   { ... }
          }

        Note: in Scalar::Does 0.007 and below the single-argument form of
        `does` returned a curried coderef. This was changed in Scalar::Does
        0.008.

    `overloads($scalar, $role)`
        A function `overloads` (which just checks overloading) is also
        available.

    `overloads($role)`
        Called with a single argument, tests $_. Yes, this works with lexical
        $_.

        Note: in Scalar::Does 0.007 and below the single-argument form of
        `overloads` returned a curried coderef. This was changed in
        Scalar::Does 0.008.

    `blessed($scalar)`, `reftype($scalar)`, `looks_like_number($scalar)`
        For convenience, this module can also re-export these functions from
        Scalar::Util. `looks_like_number` is generally more useful than
        `does($scalar, q[0+])`.

    `make_role $name, where { BLOCK }`
        Returns an anonymous role object which can be used as a parameter to
        `does`. The block is arbitrary code which should check whether $_[0]
        does the role.

    `where { BLOCK }`
        Syntactic sugar for `make_role`. Compatible with the `where` function
        from Moose::Util::TypeConstraints, so don't worry about conflicts.

  Constants
    The following constants may be exported for convenience:

    `SCALAR`
    `ARRAY`
    `HASH`
    `CODE`
    `GLOB`
    `REF`
    `LVALUE`
    `IO`
    `VSTRING`
    `FORMAT`
    `REGEXP`
    `BOOLEAN`
    `STRING`
    `NUMBER`
    `SMARTMATCH`

  Export
    By default, only `does` is exported. This module uses Exporter::Tiny, so
    functions can be renamed:

      use Scalar::Does does => { -as => 'performs_role' };

    Scalar::Does also plays some tricks with namespace::clean to ensure that
    any functions it exports to your namespace are cleaned up when you're
    finished with them. This ensures that if you're writing object-oriented
    code `does` and `overloads` will not be left hanging around as methods of
    your classes. Moose::Object provides a `does` method, and you should be
    able to use Scalar::Does without interfering with that.

    You can import the constants (plus `does`) using:

      use Scalar::Does -constants;

    The `make_role` and `where` functions can be exported like this:

      use Scalar::Does -make;

    Or list specific functions/constants that you wish to import:

      use Scalar::Does qw( does ARRAY HASH STRING NUMBER );

  Custom Role Checks
      use Scalar::Does
        custom => { -as => 'does_array', -role => 'ARRAY' },
        custom => { -as => 'does_hash',  -role => 'HASH'  };
  
      does_array($thing);
      does_hash($thing);

BUGS
    Please report any bugs to
    <http://rt.cpan.org/Dist/Display.html?Queue=Scalar-Does>.

SEE ALSO
    Scalar::Util.

    <http://perldoc.perl.org/5.10.0/perltodo.html#A-does()-built-in>.

  Relationship to Moose roles
    Scalar::Does is not dependent on Moose, and its role-checking is not
    specific to Moose's idea of roles, but it does work well with Moose roles.

    Moose::Object overrides `DOES`, so Moose objects and Moose roles should
    "just work" with Scalar::Does.

      {
        package Transport;
        use Moose::Role;
      }
  
      {
        package Train;
        use Moose;
        with qw(Transport);
      }
  
      my $thomas = Train->new;
      does($thomas, 'Train');          # true
      does($thomas, 'Transport');      # true
      does($thomas, Transport->meta);  # not yet supported!

    Mouse::Object should be compatible enough to work as well.

    See also: Moose::Role, Moose::Object, UNIVERSAL.

  Relationship to Moose type constraints
    Moose::Meta::TypeConstraint objects, plus the constants exported by
    MooseX::Types libraries all provide a `check` method, so again, should
    "just work" with Scalar::Does. Type constraint strings are not supported
    however.

      use Moose::Util::TypeConstraints qw(find_type_constraint);
      use MooseX::Types qw(Int);
      use Scalar::Does qw(does);
  
      my $int = find_type_constraint("Int");
  
      does( "123", $int );     # true
      does( "123", Int );      # true
      does( "123", "Int" );    # false

    Mouse::Meta::TypeConstraints and MouseX::Types should be compatible enough
    to work as well.

    See also: Moose::Meta::TypeConstraint, Moose::Util::TypeConstraints,
    MooseX::Types, Scalar::Does::MooseTypes.

  Relationship to Type::Tiny type constraints
    Types built with Type::Tiny and Type::Library can be used exactly as Moose
    type constraint objects above.

      use Types::Standard qw(Int);
      use Scalar::Does qw(does);
  
      does(123, Int);   # true

    In fact, Type::Tiny and related libraries are used extensively in the
    internals of Scalar::Does 0.200+.

    See also: Type::Tiny, Types::Standard.

  Relationship to Role::Tiny and Moo roles
    Roles using Role::Tiny 1.002000 and above provide a `DOES` method, so
    should work with Scalar::Does just like Moose roles. Prior to that
    release, Role::Tiny did not provide `DOES`.

    Moo's role system is based on Role::Tiny.

    See also: Role::Tiny, Moo::Role.

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENCE
    This software is copyright (c) 2012-2014 by Toby Inkster.

    This is free software; you can redistribute it and/or modify it under the
    same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
Scalar-Does
===========

Created:      2012-08-01
Home page:    <https://metacpan.org/release/Scalar-Does>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=Scalar-Does>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

0.202	2014-04-05

 [ Bug Fixes ]
 - Make t/iod-02filehandles.t clean up after itself.
   Fixes RT#91497.
   Jon Jensen++
   Piers Cawley++
   <https://rt.cpan.org/Ticket/Display.html?id=91497>
   <https://github.com/tobyink/p5-scalar-does/pull/1>

0.201	2013-09-30

 [ Packaging ]
 - Use Dist::Inkt.

 [ Other ]
 - Updated: Use Exporter::Tiny instead of Exporter::TypeTiny.

0.200	2013-05-24

 [ Documentation ]
 - Document relationship with Type::Tiny.

 [ Packaging ]
 - Skip tests involving given/when, smart match, and lexical $_ in Perl >=
   5.17.

 [ Other ]
 - Updated: Deprecate Scalar::Does::MooseTypes.
 - Updated: Use Exporter::TypeTiny instead of Sub::Exporter.
 - Updated: Use Type::Tiny and Types::Standard to perform type checking for
   built-in Scalar::Does roles, and also Scalar::Does::MooseTypes roles.

0.102	2013-03-10

 [ Bug Fixes ]
 - Scalar::Util version in lib/Scalar/Does.pm and meta/makefile.pret should
   match.

0.101	2013-03-08

 [ Documentation ]
 - Show how IO::Detect's constants can be used with Scalar::Does's does()
   function.

 [ Other ]
 - IO::Detect now performs the same namespace::clean trickery as
   Scalar::Does to fake lexical subs.
 - Improve test suite cleanliness; use Test::Requires, etc.
 - does($class, $overloaded_operation) no longer returns true.
   does($instance, $overloaded_operation) still works of course!
 - use lexical::underscore instead of directly mucking around with
   PadWalker.

0.100	2013-03-07

 [ Packaging ]
 - Merge IO-Detect distribution into Scalar-Does.
 - The IO-Detect-Changes.txt file lists changes to the IO-Detect
   distribution prior to merge.

 [ Other ]
 - Updated: New test case for IO::Detect/Path::Tiny interaction.

0.012	2013-01-11

 [ Documentation ]
 - Copyright 2013.

0.011_01	2012-12-15

 [ Bug Fixes ]
 - Attempt to fix some problems CPAN smoke tests are reporting in
   09warnings.t.

0.010	2012-12-05

 [ Documentation ]
 - Add documentation for test suite.

 [ Other ]
 - Updated: Perl 5.17 has changed the behaviour of UNIVERSAL::can($pkg,
   'can'), so that it is no longer useful for detecting whether a package
   is loaded. Changed Scalar::Does::MooseLike to avoid this idiom.
   Fixes RT#81622.
   DAMI++
   <https://rt.cpan.org/Ticket/Display.html?id=81622>

0.009	2012-11-05	Gunpowder, treason and plot

 [ Bug Fixes ]
 - Skip test '10underscore.t' on Perl 5.008 as it doesn't support lexical
   $_.
 - use if $] < 5.010, 'UNIVERSAL::DOES'.

 [ Packaging ]
 - Requires UNIVERSAL::DOES in Perl < 5.010.

0.008	2012-10-27

 [ Documentation ]
 - Remove references to Object-DOES in documentation.
 - Revise documentation about Role::Tiny/Moo::Role's lack of DOES; it is
   now implemented.

 [ Other ]
 - INCOMPATIBLE CHANGE: single-argument does($role) and overloads($role)
   now check $_.
   DAMI++
   <http://purl.org/NET/cpan-uri/rt/ticket/80434>

0.007	2012-10-23	Skyfall

 [ Bug Fixes ]
 - Fix warning on does(undef, $role) when $^W is true.
   Fixes RT#80121.
   DAMI++
   <https://rt.cpan.org/Ticket/Display.html?id=80121>

0.006	2012-10-07

 [ Packaging ]
 - Fix failing test case in Perl 5.8.

0.005	2012-10-05

 [ Documentation ]
 - Correct inaccurate information about Moo type constraints.

0.004	2012-10-03

 [ Documentation ]
 - Better document interaction with Role::Tiny and Moo.

 [ Other ]
 - Added: Scalar::Does::MooseTypes.
 - Added: make_role function for creating anonymous roles.

0.003	2012-09-19

 [ Bug Fixes ]
 - Ensure t/99smartmatch.t is skipped properly on Perl < 5.10.1.

0.002	2012-09-19

 [ Bug Fixes ]
 - Correctly determine Regexp role.

 [ Documentation ]
 - Better documentation for interaction with Moose roles and type
   constraints.
 - Clarify that does($thing, '0+') is not the same as looks_like_number.
 - Document relationship with Object::DOES.
 - Minor improvements.

 [ Packaging ]
 - Improved test suite with 100% Devel::Cover coverage.

 [ Other ]
 - Added: Export constants for pre-defined roles.
 - Added: Export looks_like_number.

0.001	2012-09-02	Initial release


Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"       => "like ref() but useful",
  "author"         => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>"],
  "dynamic_config" => 1,
  "generated_by"   => "Dist::Inkt::Profile::TOBYINK version 0.015, CPAN::Meta::Converter version 2.120921",
  "keywords"       => [],
  "license"        => ["perl_5"],
  "meta-spec"      => {
                        url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                        version => 2,
                      },
  "name"           => "Scalar-Does",
  "no_index"       => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"        => {
                        configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                        runtime => {
                          requires => {
                            "Exporter::Tiny"      => 0.026,
                            "lexical::underscore" => 0,
                            "namespace::clean"    => 0.19,
                            "Scalar::Util"        => 1.24,
                            "Type::Tiny"          => 0.004,
                            "URI::file"           => 0,
                          },
                        },
                        test => {
                          requires => {
                            "Test::More" => 0.96,
                            "Test::NoWarnings" => 0,
                            "Test::Requires" => 0.06,
                            "URI" => 0,
                          },
                        },
                      },
  "provides"       => {
                        "IO::Detect" => { file => "lib/IO/Detect.pm", version => 0.202 },
                        "IO::Detect::SmartMatcher" => { file => "lib/IO/Detect.pm", version => 0.202 },
                        "Scalar::Does" => { file => "lib/Scalar/Does.pm", version => 0.202 },
                        "Scalar::Does::MooseTypes" => { file => "lib/Scalar/Does/MooseTypes.pm", version => 0.202 },
                        "Scalar::Does::RoleChecker" => { file => "lib/Scalar/Does.pm", version => 0.202 },
                      },
  "release_status" => "stable",
  "resources"      => {
                        bugtracker   => { web => "http://rt.cpan.org/Dist/Display.html?Queue=Scalar-Does" },
                        homepage     => "https://metacpan.org/release/Scalar-Does",
                        license      => ["http://dev.perl.org/licenses/"],
                        repository   => {
                                          type => "git",
                                          url  => "git://github.com/tobyink/p5-scalar-does.git",
                                          web  => "https://github.com/tobyink/p5-scalar-does",
                                        },
                        X_identifier => "http://purl.org/NET/cpan-uri/dist/Scalar-Does/project",
                      },
  "version"        => 0.202,
  "x_contributors" => ["Piers Cawley (PDCAWLEY) <pdcawley\@cpan.org>"],
};

my %dynamic_config;
do {
$meta->{prereqs}{runtime}{requires}{'UNIVERSAL::DOES'} = 0 if $] < 5.010000;

};

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	test       => { TESTS => "t/*.t" },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

